### ISSUE TITLE
Add support for read/write timeouts

### ISSUE THREAD

**Issue by davidmc24** [ratpack/ratpack#612]
Desired behavior:
- The ability to configure a read timeout such that, if while waiting for a request, the timeout elapses without reading any data, the connection is closed
- The ability to configure a write timeout such that, if while processing a request, the timeout elapses without writing any data, the connection is closed

Proposed design:
- Add support for getting `readTimeout` and `writeTimeout` in `ServerConfig` (and associated builder impl)
- Add support for the new settings in `ServerConfigDeserializer`
- If a `readTimeout` is configured
  - add a `ReadTimeoutHandler` to the `ChannelPipeline` (see `NettyRatpackServer#buildChannel`)
  - upon receiving a request, remove the read timeout handler (see `NettyHandlerAdapter`)
  - upon writing a response, add the read timeout handler again
- If a `writeTimeout` is configured
  - upon receiving a request, add a `WriteTimeoutHandler`
  - upon writing a response, remove the write timeout handler

**Comment by davidmc24** [comment]
Requested on the forums: http://forum.ratpack.io/How-can-I-set-timeout-for-request-td997.html

**Comment by kakumara** [comment]
Hello,
I understand this feature is not yet implemented. I'm just wondering if there is a workaround that I can do until this is implemented, handling a request timeout scenario as below.

Basically I have a scenario that needs handling a possible long running request and I need to respond either a `200` or a `202` depending on a certain time threshold is reached. In case of threshold reached, i still need to continue the transaction but respond `202` immediately to client. I tried doing this converting the ratpack promise to rx.Observable's and using timeout / timers but it leads to a bit of a convoluted mess and going nowhere. 
I want this behaviour only at this resource level not globally.

Any suggestions?

**Comment by alkemist** [comment]
It's possible to do. It would work something like this …

1. Accept the request
2. Create a `Promised` for the outcome and an `AtomicReference` to a state enum representing the outcome, set the initial state of the reference to `PENDING` 
3. Fork a new execution to do the actual work, when it completes do a state.compareAndSet(PENDING, COMPLETE) - if this succeeds, fulfil the Promised
4. After forking #3, fork a new execution as the timer and use Promise.defer() and ExecController.getExecutor() to schedule resuming the promise with the timeout value - when it resumes, do a state.compareAndSet(PENDING, TIMEOUT) and fulfil the promise if it succeeds

There's a few variations on that possible, but that's roughly how I'd approach it. If you would like some further help, please jump on the Slack channel and we can discuss (details on ratpack.io).

At some point, we'll provide sugar for exactly this kind of thing.

**Comment by kakumara** [comment]
Thanks @alkemist for your response.

After a bit of struggle I managed to get this working with 2 Rx-Observables. One doing the long running transaction and the other triggering a timer event. And by `merging` the two together I managed to respond accordingly depending on which ever occurs first.  
I'm interested to try out your suggestion too, just out of curiosity. I might actually do that and let you know how it goes.
